runelite-client/src/main/java/net/runelite/client/plugins/raids/bats/BatsLocator.java

```diff
@@ -0,0 +1,503 @@
+/*
+ * Copyright (c) 2020, chestnut1693 <chestnut1693@gmail.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.plugins.raids.bats;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.function.Function;
+import lombok.Getter;
+import net.runelite.api.Client;
+import static net.runelite.api.Constants.CHUNK_SIZE;
+import net.runelite.api.GameObject;
+import net.runelite.api.ObjectID;
+import net.runelite.api.coords.WorldPoint;
+
+public class BatsLocator
+{
+	public static final int TROUGH = ObjectID.TROUGH_29746;
+	public static final int CLOSED = ObjectID.CHEST_29742;
+	public static final int OPENED_POISON_OR_BATS = ObjectID.CHEST_29743;
+	public static final int OPENED_WITHOUT_GRUBS = ObjectID.CHEST_29744;
+	public static final int OPENED_WITH_GRUBS = ObjectID.CHEST_29745;
+
+	@Getter
+	private RoomType roomType;
+
+	@Getter
+	private HashMap<WorldPoint, Chest> chests = new HashMap<>();
+
+	@Getter
+	private HashSet<Chest> poisonBatsChests = new HashSet<>();
+
+	@Getter
+	private HashSet<Chest> grubsChests = new HashSet<>();
+
+	@Getter
+	private ArrayList<ArrayList<Chest>> solutionSets = new ArrayList<>();
+
+	@Getter
+	private int highestSolutionSetCount = 0;
+
+	@Getter
+	private boolean drawChestStates = false;
+
+	private Client client;
+	private int rotation = -1;
+
+	public BatsLocator(Client client)
+	{
+		this.client = client;
+	}
+
+	public void troughSpawnEvent(GameObject trough)
+	{
+		if (roomType == null && rotation == -1)
+		{
+			int plane = trough.getPlane();
```

Comment:
This is all just duplicated from `WorldPoint.fromLocalInstance`, use it instead.

Reply:
I'm not interested in the world point but in the variable `instanceX` as this uniquely identifies the room. The world point of the trough can be different each time the raid chambers are entered but `instanceX` is always the same. If i remember correctly, the variable `instanceX` is the base of a chunk where the trough spawns in. I've seen this approach used elsewhere.

Reply:
What you call `instanceX` is simply the base of the template chunk in world coords. So it would just be `worldPoint.getX() & -CHUNK_SIZE`

Reply:
I now see the difference between the world coord of the trough and the world coord returned from `WorldPoint.fromLocalInstance`. The variable I call `instanceX` is called `templateChunkX` in the function. I could remove all the duplicate stuff but I how do I get the rotation of the chunk then?

You find this more appropriate?
`LocalPoint localPoint = trough.getLocalLocation();`
`rotation = client.getInstanceTemplateChunks()[trough.getPlane()][localPoint.getSceneX() / CHUNK_SIZE][localPoint.getSceneY() / CHUNK_SIZE] >> 1 & 0x3;`
`switch (WorldPoint.fromLocalInstance(client, localPoint).getX() & -CHUNK_SIZE) {}`